src/librustc_mir/interpret/intern.rs

```diff
@@ -0,0 +1,267 @@
+                            }
+                            self.intern(vtable)?;
+                        }
+                    }
+                }
+            }
+            let mplace = self.ecx.ref_to_mplace(value)?;
+            // Check if we have encountered this pointer+layout combination before.
+            // Only recurse for allocation-backed pointers.
+            if let Scalar::Ptr(ptr) = mplace.ptr {
+                match self.intern(ptr)? {
+                    // No need to recurse, these are interned already and statics may have
+                    // cycles, so we don't want to recurse there
+                    Some(IsStaticOrFn) => {},
+                    // intern everything referenced by this value
+                    None => self.ref_tracking.track(mplace, || ()),
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+pub fn intern_const_alloc_recursive(
+    ecx: &mut CompileTimeEvalContext<'a, 'mir, 'tcx>,
+    def_id: DefId,
+    ret: MPlaceTy<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+) -> EvalResult<'tcx> {
+    let tcx = ecx.tcx;
+    let frozen = ret.layout.ty.is_freeze(tcx.tcx, param_env, tcx.span);
+    let intern_mode = if tcx.is_static(def_id) {
+        if tcx.is_mutable_static(def_id) {
+            InternMode::StaticMut
+        } else {
+            InternMode::Static { frozen }
+        }
+    } else {
+        InternMode::Const { frozen }
+    };
+
+    // type based interning
+
+    let mut ref_tracking = RefTracking::new(ret);
+    let leftover_relocations = &mut FxHashSet::default();
+
+    // start with the outer most allocation
+    InternVisitor {
+        ref_tracking: &mut ref_tracking,
+        ecx,
+        mode: intern_mode,
+        leftover_relocations,
+    }.intern(ret.ptr.to_ptr()?)?;
+
+    while let Some((mplace, ())) = ref_tracking.todo.pop() {
+        let interned = InternVisitor {
+            ref_tracking: &mut ref_tracking,
+            ecx,
+            mode: intern_mode,
+            leftover_relocations,
+        }.visit_value(mplace);
+        if let Err(error) = interned {
+            // This can happen when e.g. the tag of an enum is not a valid discriminant. We do have
+            // to read enum discriminants in order to find references in enum variant fields.
+            if let InterpError::ValidationFailure(_) = error.kind {
+                let err = crate::const_eval::error_to_const_error(&ecx, error);
+                match err.struct_error(ecx.tcx, "it is undefined behavior to use this value") {
+                    Ok(mut diag) => {
+                        diag.note("The rules on what exactly is undefined behavior aren't clear, \
+                            so this check might be overzealous. Please open an issue on the rust \
+                            compiler repository if you believe it should not be considered \
+                            undefined behavior",
```

Comment:
Would be nice to deduplicate this (we have [similar text elsewhere](https://github.com/rust-lang/rust/blob/6cdaba34fc0bb36a306800d535c0feafa1afada2/src/librustc_mir/const_eval.rs#L552)). But that can also happen in a separate PR.